### PR TITLE
Location

### PR DESCRIPTION
**Pull request by dsekulovic**

### CODE REVIEW

File: src/app/app.component.html [hunk 1]
```diff
@@ -10,7 +10,7 @@
           <a class="nav-link" routerLink="/">Home</a>
         </li>
         <li class="nav-item" routerLinkActive="active">
-          <a class="nav-link" routerLink="/location">Countries</a>
+          <a class="nav-link" routerLink="/location/serbia">Countries</a>
```
Comment (aberonja) [inline src/app/app.component.html]: Why is this called countries if it goes to 🇷🇸 route?

File: src/app/charts/charts.component.html [hunk 2]
```diff
@@ -1,18 +1,16 @@
...
     <div class="btn-group btn-group-sm btn-group-toggle" data-toggle="buttons">
-        <label class="btn" [ngClass]="{
-            'btn-danger active':label === 'TotalConfirmed' || label === 'NewConfirmed',
-            'btn-success':label === 'TotalRecovered' || label === 'NewRecovered',
-            'btn-dark':label === 'TotalDeaths' || label === 'NewDeaths'
-          }" *ngFor="let label of labels">
-            <input type="radio" name="options" (click)="onClick(label)">
-            {{ label }}
+        <label class="btn btn-outline-secondary" [ngClass]="{
+            'active':info === 'TotalConfirmed' || info === 'NewConfirmed' || info === 0
```
Comment (aberonja) [inline src/app/charts/charts.component.html]: [class.active]="isClassActive()"

File: src/app/charts/charts.component.html [hunk 3]
```diff
@@ -1,18 +1,16 @@
...
+          }" *ngFor="let info of (period || labels)">
+            <input type="radio" name="options" (click)="onClick(info)">
+            {{ period ? info === 0 ? 'Whole period' : 'Last ' + info + ' days' : info }}
```
Comment (aberonja) [inline src/app/charts/charts.component.html]: Not the best practice to have a ternary inside of a ternary.

File: src/app/charts/charts.component.ts [hunk 4]
```diff
@@ -31,38 +40,56 @@ export class ChartsComponent implements OnInit {
...
     this.lineChartData = this.labels.reduce(
       (acc, el) => [...acc, { data: [], label: el }],
       []
     );
+
     this.loadChartData(this.labels[0]);
   }
 
-  onClick(type: string) {
+  ngOnChanges(changes: SimpleChanges) {
```
Comment (aberonja) [inline src/app/charts/charts.component.ts]: Could this be onInit? If the route is changed this component should be reinitialized, right?

Also, you don't need id variable.

File: src/app/http.service.ts [hunk 5]
```diff
@@ -10,12 +10,53 @@ export class HttpClass {
...
+  getAllCountries() {
+    return this.http.get<any>("https://api.covid19api.com/countries");
+  }
+
+  getDataForCountry(slug: string) {
+    return this.http
+      .get<any>(`https://api.covid19api.com/dayone/country/${slug}`)
+      .pipe(
+        map((data) => {
+          const newData = [];
```
Comment (aberonja) [inline src/app/http.service.ts]: \n

File: src/app/http.service.ts [hunk 6]
```diff
@@ -10,12 +10,53 @@ export class HttpClass {
...
+          for (const el of data) {
```
Comment (aberonja) [inline src/app/http.service.ts]: Why for? Please use .map everywhere from now on 😄

File: src/app/http.service.ts [hunk 7]
```diff
@@ -10,12 +10,53 @@ export class HttpClass {
...
+            newData.push({
+              ...el,
+              Date: new Date(el.Date)
+                .toDateString()
+                .split(" ")
+                .slice(1)
+                .join(" "),
+              // .toLocaleDateString(),
```
Comment (aberonja) [inline src/app/http.service.ts]: Please remove this.

File: src/app/http.service.ts [hunk 8]
```diff
@@ -10,12 +10,53 @@ export class HttpClass {
...
+            });
+          }
```
Comment (aberonja) [inline src/app/http.service.ts]: \n

File: src/app/location/filter.pipe.ts [hunk 9]
```diff
@@ -0,0 +1,15 @@
+import { Pipe, PipeTransform } from "@angular/core";
+import { ICountryInfo } from "../interface/interface";
+
+@Pipe({
+  name: "filter",
+})
+export class FilterPipe implements PipeTransform {
+  transform(items: ICountryInfo[], searchText: string): any[] {
+    if (!items) return [];
```
Comment (aberonja) [inline src/app/location/filter.pipe.ts:9]: \n

File: src/app/location/filter.pipe.ts [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+    if (!searchText) return items;
```
Comment (aberonja) [inline src/app/location/filter.pipe.ts:11]: \n

File: src/app/location/filter.pipe.ts [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+    return items.filter((it) =>
```
Comment (aberonja) [inline src/app/location/filter.pipe.ts]: When creating temp variable, please give it meaningful name, like **item**.

File: src/app/location/location.component.html [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+        <div class="map d-flex" *ngIf="!error">
+            <agm-map [latitude]="latitude" [longitude]="longitude">
+                <agm-marker [latitude]="latitude" [longitude]="longitude"></agm-marker>
+            </agm-map>
+        </div>
+        <div class="d-xl-flex flex-xl-row-reverse p-3" *ngIf="!error">
+            <div class="col-xl-5 bg-light text-center pt-5 font-weight-bold align-self-center" style="height: 300px">
+                <h1>{{ name }}</h1>
+                <h3>{{ temp }}<sup>&#x2103;</sup></h3>
+                <h4 class="mt-5">{{temp>25 ? 'We are safe!': 'Corona will kill us!'}}</h4>
```
Comment (aberonja) [inline src/app/location/location.component.html]: 🦀

File: src/app/location/location.component.scss [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+  max-height: 200px;
+  overflow-y: auto;
+  margin-bottom: 10px;
+  border: 1px solid gray;
+}
+
+.map {
+  width: 100%;
+  min-height: 40vh;
+  height: 0;
```
Comment (aberonja) [inline src/app/location/location.component.scss]: 😕 ❓

File: src/app/location/location.component.scss [hunk 14]
```diff
@@ -0,0 +1,39 @@
+.countries-list {
+  max-height: 200px;
+  overflow-y: auto;
+  margin-bottom: 10px;
+  border: 1px solid gray;
```
Comment (aberonja) [inline src/app/location/location.component.scss]: These could be bootstrap classes, right?

File: src/app/location/location.component.scss [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+.map {
+  width: 100%;
+  min-height: 40vh;
+  height: 0;
+
+  agm-map {
+    height: 100%;
+    width: 100%;
```
Comment (aberonja) [inline src/app/location/location.component.scss]: Bootstrap classes.

File: src/app/location/location.component.scss [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+    width: 200px;
+  }
+
+  .countries-list {
+    max-height: calc(100vh - 120px);
+    overflow-y: auto;
+  }
+
+  .side-data {
+    height: calc(100vh - 58px);
```
Comment (aberonja) [inline src/app/location/location.component.scss]: This looks odd, please explain.

File: src/app/location/location.component.ts [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+  ngOnInit(): void {
+    this.subs.push(
+      this.http.getAllCountries().subscribe((data) => {
+        this.countries = data.sort((a: ICountryInfo, b: ICountryInfo) =>
+          a.Country > b.Country ? 1 : -1
+        );
+      })
+    );
+
+    let id = "sebia";
```
Comment (aberonja) [inline src/app/location/location.component.ts]: 🇷🇸 ❓

Comment (aberonja) [reply src/app/location/location.component.ts]: Could this be const?

File: src/app/location/location.component.ts [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+          this.error = null;
+        } else {
+          this.error = "No data!";
+          this.isLoading = false;
+        }
+      })
+    );
+  }
+
+  ngOnDestroy() {
```
Comment (aberonja) [inline src/app/location/location.component.ts]: Component order:

@input @output
Properties
Constructor
Lifecycle methods
Custom methods

File: src/app/location/location.component.ts [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+        );
+      })
+    );
+
+    let id = "sebia";
+    this.route.params.subscribe((params) => (id = params["id"]));
+    this.dataForCountry(id);
+  }
+
+  onClick(data: string) {
```
Comment (aberonja) [inline src/app/location/location.component.ts]: onClick naming is not a good one, what does this method do?

File: src/app/location/location.component.ts [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+        if (data.length) {
+          this.latitude = +data[0].Lat;
+          this.longitude = +data[0].Lon;
+
+          this.subs.push(
+            this.http
+              .getWeather(this.latitude, this.longitude)
+              .subscribe(
+                (data) => (
+                  (this.temp = Math.round(data.temp)),
```
Comment (aberonja) [inline src/app/location/location.component.ts]: This could be an object instead of 3 separate props.

**Review by aberonja — CHANGES_REQUESTED** [review]
👷

File: src/app/app.component.html (hunk 1 above)
Comment (dsekulovic) [reply src/app/app.component.html]: I made 'serbia' default route. On that page (/location/:id), you have a list of countries where you can select one to show data for that country

Comment (aberonja) [reply src/app/app.component.html]: Let's call this route home country, and then we can have a dropdown or something similar with top 10 countries for example which will take us to that route. 
(You can create top 10 countries array by your choice).

Or, if you already have countries route you can have a list of countries and then clicking on one will take you to that country route by id.

File: src/app/location/location.component.scss (hunk 16 above)
Comment (dsekulovic) [reply src/app/location/location.component.scss]: That is for adding scroller for part of the page where map, chart and weather data are.
I don't want scroller for whole page, 'couse I have one for country list too, and I need them to be in sync

Comment (dsekulovic) [reply src/app/location/location.component.scss]: scroller

Comment (dsekulovic) [reply src/app/location/location.component.scss]: To avoid this...

File: src/app/location/location.component.ts (hunk 19 above)
Comment (dsekulovic) [reply src/app/location/location.component.ts]: This method displays data for country that's been clicked (triggers this.dataForCountry(data))

File: src/app/charts/charts.component.html [hunk 21]
```diff
@@ -1,18 +1,15 @@
...
-            'btn-danger active':label === 'TotalConfirmed' || label === 'NewConfirmed',
-            'btn-success':label === 'TotalRecovered' || label === 'NewRecovered',
-            'btn-dark':label === 'TotalDeaths' || label === 'NewDeaths'
-          }" *ngFor="let label of labels">
-            <input type="radio" name="options" (click)="onClick(label)">
-            {{ label }}
+        <label class="btn btn-outline-secondary" [class.active]="isClassActive(info)"
+            *ngFor="let info of (period || labels)">
+            <input type="radio" name="options" (click)="onClick(info)">
+            {{ info | texttransform }}
```
Comment (aberonja) [inline src/app/charts/charts.component.html]: *textTransform

File: src/app/charts/charts.component.ts [hunk 22]
```diff
@@ -31,38 +34,59 @@ export class ChartsComponent implements OnInit {
...
+  constructor(private route: ActivatedRoute) {}
 
   ngOnInit(): void {
     this.lineChartData = this.labels.reduce(
       (acc, el) => [...acc, { data: [], label: el }],
       []
     );
-    this.loadChartData(this.labels[0]);
+
+    this.route.params.subscribe((params) =>
```
Comment (aberonja) [inline src/app/charts/charts.component.ts]: Unsubscribe is missing.

File: src/app/charts/charts.component.ts [hunk 23]
```diff
@@ -31,38 +34,59 @@ export class ChartsComponent implements OnInit {
...
+      this.loadChartData(params["id"] || this.labels[0])
+    );
   }
 
-  onClick(type: string) {
+  onClick(type: any) {
```
Comment (aberonja) [inline src/app/charts/charts.component.ts]: Call this method loadChartData() or something similar.

File: src/app/charts/charts.component.ts [hunk 24]
```diff
@@ -31,38 +34,59 @@ export class ChartsComponent implements OnInit {
...
+  loadChartData(type: any) {
     this.activeSort = type;
+    let filteredData = [];
 
-    const fiteredData = this.chartData
-      .sort((a: ICountryInfo, b: ICountryInfo) => (a[type] < b[type] ? 1 : -1))
-      .slice(0, this.numberOfData);
+    if (this.period) {
+      filteredData = this.chartData.slice(0 - type);
+      this.text = `Covid-19 in ${this.chartData[0].Country}`;
```
Comment (aberonja) [inline src/app/charts/charts.component.ts:66]: \n

File: src/app/charts/charts.component.ts [hunk 25]
```diff
@@ -31,38 +34,59 @@ export class ChartsComponent implements OnInit {
...
       []
     );
 
-    this.lineChartLabels = fiteredData.map((el) => el.Country);
+    this.lineChartLabels = filteredData.map((el) =>
+      this.period ? el.Date : el.Country
+    );
+  }
+
+  isClassActive(data) {
```
Comment (aberonja) [inline src/app/charts/charts.component.ts]: This is simple enough to be in the template directly. Not a mistake, but keep it in mind. 🧄

File: src/app/charts/text-transform.pipe.ts [hunk 26]
```diff
@@ -0,0 +1,14 @@
+import { Pipe, PipeTransform } from "@angular/core";
+
+@Pipe({
+  name: "texttransform",
+})
+export class TextTransformPipe implements PipeTransform {
+  transform(data: any): string {
+    if (typeof data === "number") {
+      return data ? `Last ${data} days` : "Whole period";
+    }
```
Comment (aberonja) [inline src/app/charts/text-transform.pipe.ts:10]: \n

File: src/app/charts/text-transform.pipe.ts [hunk 27]
```diff
@@ -0,0 +1,14 @@
...
+    if (data.includes("Total")) return data.replace("Total", "Total ");
```
Comment (aberonja) [inline src/app/charts/text-transform.pipe.ts:12]: \n

File: src/app/location/location.component.scss (hunk 16 above)
Comment (aberonja) [reply src/app/location/location.component.scss]: Could this be done with flex-grow?

File: src/app/location/location.component.ts [hunk 28]
```diff
@@ -0,0 +1,85 @@
...
+  ngOnInit(): void {
+    this.subs.push(
+      this.http.getAllCountries().subscribe((data) => {
+        this.countries = data.sort((a: ICountryInfo, b: ICountryInfo) =>
+          a.Country > b.Country ? 1 : -1
+        );
+      })
+    );
+
+    this.route.params.subscribe((params) => this.dataForCountry(params["id"]));
```
Comment (aberonja) [inline src/app/location/location.component.ts]: Unsubscribe is missing.

**Review by aberonja — CHANGES_REQUESTED** [review]
Just a bit more 🧷

File: src/app/location/location.component.scss (hunk 16 above)
Comment (dsekulovic) [reply src/app/location/location.component.scss]: If it can, I'm not making it work :(

**Review by aberonja — APPROVED** [review]

**Pull request merged**